src/decode.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::{Expect, Index, Mir, FORMAT_VERSION};
+use rmp_serde::{
+    decode::{self, ReadReader},
+    Deserializer,
+};
+use serde::Deserialize;
+use std::io::Read;
+
+/// A decoder.
+pub struct Decoder<'a> {
+    deser: Deserializer<ReadReader<&'a mut dyn Read>>,
+    index: Index,
+}
+
+impl<'a> Decoder<'a> {
+    /// Returns a new decoder which will deserialise from `read_from` and also an `Index` informing
+    /// the consumer of the expected contents.
+    pub fn new(read_from: &'a mut dyn Read) -> Result<(Index, Self), decode::Error> {
+        let mut deser = Deserializer::new(read_from);
+        let ver = usize::deserialize(&mut deser)?;
+        if ver != FORMAT_VERSION {
+            panic!("incorrect version number!");
+        }
+
+        let index = Index::deserialize(&mut deser)?;
+        Ok((
+            index.clone(),
```

Comment:
Why are we cloning the index as part of a tuple when we include it in the struct too?

Reply:
The one in the struct is used to keep track of what the deserialiser has already given out, and is mutated each time the user is given (in this case) a MIR.

This will become more important later where the deserialiser will likely give out other kinds of data than just MIR. In that case, the data would have to be read out by the consumer in the correct order.

I considered using a state machine to instead of a mutable copy of the index, but decided I was trying to be too clever.

I also experimented with having a deserialiser which acts only as an interface to hand out iterators, so that the order the user requests the data is irrelevant. The problem there is that we can only have one mutable reference to the underlying `T: Read` and once we give it to serde, it's gone forever. This means we can't seek the underlying stream for subsequent iterator requests.

I hope that made sense, it's hard to explain.